Make Lua overlays + printed messages scale with HUD size

SolraBizna opened the pull request:
Back when I first added Lua overlays, there were no HUD scaling options. When they were added, Lua overlays and "prints" were left in the past. When playing in 1080p, potentially critical script-related information became very easy to miss. This PR makes the "on-screen font" text size scale under the same circumstances as the HUD, thus making all script-driven information channels more or less equally visible at different resolutions. (People who are attached to tiny text are invited to change the on-screen font to a smaller one via MML.)

As a side note, overlays now adapt to different on-screen font settings in a sane way. Previously, many aspects of their positioning were hard-coded...

Finally, because I was an idiot when I originally wrote the code, the Lua overlays only ever displayed overlays for the local player, even when viewing another player. This PR also makes it so that overlays are actually tracked for all players, and displays whichever set of overlays is applicable for the currently-viewed player. (This is important in co-op, team games, and film playback.)

This PR should not affect film playback, saved games, or network games, and has no known unintended side effects.

Hopper262 commented:
**Scaling:** Merged as 900c832 and 2bdf82f. While I've been meaning for years to move this drawing to the HUD layer, it keeps not happening, and the lack of scaling really stands out on modern monitors. The overlays seem to be about one character closer than before, but otherwise they feel the same and fit well with the font scaling. Thanks!

**Overlay player display:** During active play (as opposed to films), switching the behavior has a few downsides, so I'm holding off for now:

 * Elimination.lua deliberately uses the current behavior to positive effect. Elimination has last-man-standing gameplay, and after you die, it transfers the view with `Player:view_player()` while showing overlay data relevant to the local player. This commit would break that expected behavior.

* `Player:print()` and `Player.texture_palette` also operate on the local player only. Overlays would become consistent with the HUD, but inconsistent with other Lua-driven displays, which is arguably worse.

* Scripters have _almost_ everything they need to implement this behavior without engine help, and choose which is appropriate. If the Lua API provided something like `Player.viewed_player`, scripts could detect the local player's current view, and populate overlays accordingly. That doesn't preclude us from adding code to make scripters' lives easier, but it is harder to justify adding engine complexity when the job can be done in Lua.

For film playback, ideally you'd effectively switch the local player each time you switch the view. Lua will always be problematic, since scripts can do twisted things like cache `Player.local_` and make display decisions based on that. This commit might still be the best way forward.

One other thing: reducing the maximum overlay text from 255 characters to 31 characters is a user-visible change in behavior, even though we encourage scripters to keep overlays short.

SolraBizna commented:
> (about overlays operating for local player only)

I would be okay with continuing to have this be the case, except for the fact that overlays are part of the Player object, and not part of global state. They were always intended to be player-specific, and the only reason they applied only to the local player in my original implementation is that I honestly forgot about viewing other players' perspectives.

I also have some existing scripts I'm fond of that rely on the current (wrong) behavior.

I would be in favor of adding a version gate system that, in general, allows us to make non-backwards-compatible changes in script behavior. The old overlay behavior would then be easy to preserve behind a version gate, and then the new behavior could be the norm for all new scripts.

Either something like starting a script with:

    SCRIPT_VERSION = 2

which would not allow us to easily change anything that would be visible before trigger calls, or:

    -- A1 lua version 2

which means we have to parse the script for it...

It might also be good to have two sets of overlays: a global set and a player-specific set. (Maybe a team-specific set too?) It might also also be good to have overlays automatically change size and wrap according to their contents. It might _also_ be good to allow an arbitrary number of overlays. All of this would have to go behind a version gate...

> One other thing: reducing the maximum overlay text from 255 characters to 31 characters is a user-visible change in behavior, even though we encourage scripters to keep overlays short.

I agree that it is problematic. In fact, I ran into a problem with it on a real script almost immediately. I originally made the change in order to reduce the memory footprint, and I never went back and changed it because I was neck deep in `60fps` and I am _really_ not comfortable with git branches...

I think it would be better to replace it with a `std::string`, in keeping with the general trend of modernizing the code. That would also allow embedded NULs, which might become significant if overlays ever come under the control of Lua HUD scripts.

I'd be happy to do the legwork on either or both of the above points if I get the go-ahead. (Also, welcome back.)

Hopper262 commented:
A versioning approach works for me. We have the precedent of `Game.proper_item_accounting` we can follow. When `Game.overlays_follow_player_view` is false (the default), the overlay for local_player_index is drawn; when true, the overlay for current_player_index is drawn. The flag is checked at draw time, so scripters can request either behavior or even change it contextually. The flag also is also checked when `Player.print()` is called, discarding either non-local or non-current messages. We can leave `Player.texture_palette` the way it is, since it's such a specialized feature.

I'd like to replace as many C strings with `std::string` as we can, so refactoring the overlay code for that is definitely welcome.

And it's good to be back. Fingers crossed my workload allows me to stick around for a while.

SolraBizna commented:
I've implemented it more or less as you asked (except that I named the variable `Game.nonlocal_overlays`) and also added `Game.view_player` and `Game.local_player` variables. I can commit my changes as soon as someone tells me how to `regenerate docs/Lua.html`. :|

Hopper262 commented:
I use `xsltproc Lua.xsl Lua.xml > Lua.html`. If xsltproc misbehaves or isn't installed, I can regen the HTML after merging. Thanks!

SolraBizna commented:
Done. Hopefully this meets with approval.

Hopper262 commented:
Life keeps yanking me away unexpectedly, but I think everything is in now. I like the `local_player` and `view_player` shortcuts, but moved them to the `Players` object to keep related things together. If I forgot something, or the API still doesn't seem to work as it should, feel free to reopen this or start a new PR/issue. Thanks again for your work, I really appreciate it!
